packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,208 @@
+    this.logger.debug("Initializing backfill sync from finalizedCheckpoint", {
+      checkpoint: ssz.phase0.Checkpoint.toJson(finalizedCheckpoint),
+    });
+    this.anchorBlock = await this.db.blockArchive.getByRoot(finalizedCheckpoint.root);
+    void this.sync();
+    this.processor.trigger();
+  }
+
+  async sync(): Promise<void> {
+    try {
+      for await (const _ of this.processor) {
+        this.logger.debug("Backfill Sync - Peer count", {peerCount: this.peers.size});
+        if (this.peers.size === 0) continue;
+
+        const oldestRequiredEpoch = Math.max(
+          GENESIS_EPOCH,
+          this.chain.clock.currentEpoch - getMinEpochForBlockRequests(this.config)
+        );
+        const oldestSlotRequired = computeStartSlotAtEpoch(oldestRequiredEpoch);
+        this.logger.debug("Backfill sync", {
+          oldestSlotRequired,
+          anchorSlot: this.anchorBlock?.message.slot ?? Infinity,
+        });
+        if (this.anchorBlock) {
+          try {
+            const toSlot = this.lastFetchedSlot ?? this.anchorBlock.message.slot;
+            const fromSlot = Math.max(toSlot - this.opts.batchSize, oldestSlotRequired);
+            if (fromSlot >= toSlot) {
+              this.emit(BackfillSyncEvent.completed, toSlot);
+              this.logger.info("Backfill sync completed");
+              break;
+            }
+            this.logger.debug("BackfillSync - range sync", {
+              fromSlot,
+              toSlot,
+              anchorSlot: this.anchorBlock.message.slot,
+              lastFetched: this.lastFetchedSlot ?? -1,
+            });
+            await this.syncRange(fromSlot, toSlot, this.anchorBlock.message.parentRoot);
+          } catch (e) {
+            this.logger.debug("Error while backfiling by range", e);
+          }
+        } else {
+          try {
+            await this.syncBlock(this.chain.getFinalizedCheckpoint().root);
+          } catch (e) {
+            this.logger.debug("Error while backfilling anchor block", e);
+          }
+        }
+      }
+    } catch (e) {
+      if (e instanceof ErrorAborted) {
+        return; // Ignore
+      }
+      this.logger.error("BackfillSync Error", e);
+    }
+  }
+
+  private addPeer = (peerId: PeerId, peerStatus: phase0.Status): void => {
+    const requiredSlot =
+      this.anchorBlock?.message.slot ?? computeStartSlotAtEpoch(this.chain.getFinalizedCheckpoint().epoch);
+    this.logger.debug("Backfill sync - add peer", {peerhead: peerStatus.headSlot, requiredSlot});
+    if (peerStatus.headSlot >= requiredSlot) {
+      this.peers.set(peerId, peerStatus);
+      this.processor.trigger();
+    }
+  };
+
+  /**
+   * Remove this peer from all sync chains
+   */
+  private removePeer(peerId: PeerId): void {
+    this.network.peerRpcScores.applyAction(peerId, PeerAction.MidToleranceError);
+    this.peers.delete(peerId);
+  }
+
+  private async syncBlock(root: Root): Promise<void> {
+    const peer = getRandomPeer(this.peers.keys());
```

Comment:
I think we have ways of ranking peers by score?

Reply:
don't think we do that even in range sync. We use score to determine when to disconnect peer which will remove it from this set as well.